### PR TITLE
[SILGen] Add flag to skip typechecking and SIL gen for all function bodies

### PR DESCRIPTION
**Pull request by bnbarham**
Adds a new flag "-experimental-skip-all-function-bodies" that skips SIL
generation entirely. Function bodies will not be type checked, but may
still be parsed if another flag forces delayed parsing off.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

File: include/swift/Basic/LangOptions.h [hunk 1]
```diff
@@ -493,6 +493,10 @@ namespace swift {
     /// function bodies.
     bool SkipNonInlinableFunctionBodies = false;
 
+    /// Indicate that the type checker should skip type-checking all
+    /// function bodies, where not otherwise required for type inference.
+    bool SkipAllFunctionBodies = false;
```
Comment (harlanhaskins) [inline include/swift/Basic/LangOptions.h:498]: Would it make sense to scoop these two bools into an enum?

```
enum class FunctionBodySkipping : uint8_t {
  None,
  NonInlinable,
  All
}
```

Comment (bnbarham) [reply include/swift/Basic/LangOptions.h:498]: As far as I understand it, the idea is that `SkipNonInlinableFunctionBodies` will actually be the default at some point, with no CLI option. So making it an enum here would make that harder to do in the future. However if that's not true, I'm happy to change it to an enum instead.

Comment (slavapestov) [reply include/swift/Basic/LangOptions.h:498]: I think we only ever want to skip non-inlinable function bodies when building a swiftinterface file or TBD. Consolidating these flags into an enum makes sense to me.

**Comment by slavapestov** [comment]
I think these days we don't have a flag for delayed parsing anymore, we just always lazily parse function bodies in non-primary files the first time you call AbstractFunctionDecl::getBody(). So if nothing asks for the body of a function it won't be parsed.

**Comment by akyrtzi** [comment]
The test seems to only focus on checking SIL generation but there doesn't seem to be a test to check that typechecking is really avoided.
I'd suggest to add some test cases to check for that, see uses of `-Xfrontend -debug-forbid-typecheck-prefix` in other tests to see how you could accomplish this.

**Comment by bnbarham** [comment]
> I think these days we don't have a flag for delayed parsing anymore, we just always lazily parse function bodies in non-primary files the first time you call AbstractFunctionDecl::getBody(). So if nothing asks for the body of a function it won't be parsed.

Sorry, may not have been clear in my description there. By "another flag" I was mostly referring to eg. "-dump-parse", "-verify-syntax-tree", etc. Basically, all the conditions in `hasDelayedBodyParsing`.

**Comment by bnbarham** [comment]
> The test seems to only focus on checking SIL generation but there doesn't seem to be a test to check that typechecking is really avoided.
> I'd suggest to add some test cases to check for that, see uses of `-Xfrontend -debug-forbid-typecheck-prefix` in other tests to see how you could accomplish this.

Will do, thanks.

File: include/swift/Basic/LangOptions.h (hunk 1 above)
Comment (bnbarham) [reply include/swift/Basic/LangOptions.h:498]: Sorry, good point. I was just thinking about the flag itself. I'll add the enum, thanks :)

**Comment by bnbarham** [comment]
@swift-ci please test

File: include/swift/Basic/LangOptions.h (hunk 1 above)
Comment (bnbarham) [reply include/swift/Basic/LangOptions.h]: I've put the enum under swift in LangOptions.h for now and I'm currently importing it in SILOptions.h. Does that seem okay @harlanhaskins? Alternatively, would you prefer either:
1. a separate header with just that enum, or
2. another enum defined in SILOptions.h?

Comment (harlanhaskins) [reply include/swift/Basic/LangOptions.h]: Another enum in SILOptions.h is fine, thank you!

**Review by akyrtzi — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test

File: include/swift/Basic/LangOptions.h (hunk 1 above)
Comment (bnbarham) [reply include/swift/Basic/LangOptions.h]: Spoke with Harlan privately - ended up deciding on a separate header. Thanks :)!

File: test/Frontend/skip-function-bodies.swift [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+// ERROR: -experimental-skip-*-function-bodies do not support emitting IR
+
+// Warn when trying to build SwiftONoneSupport with skip non-inlinable
+// function bodies enabled
+// RUN: %target-swift-frontend -typecheck -experimental-skip-non-inlinable-function-bodies -module-name SwiftOnoneSupport %s 2>&1 | %FileCheck %s --check-prefix WARNING
+// WARNING: module 'SwiftOnoneSupport' cannot be built with -experimental-skip-non-inlinable-function-bodies; this option has been automatically disabled
+
+// Check skipped bodies are neither typechecked nor SILgen'd
+// RUN: %target-swift-frontend -emit-sil -emit-sorted-sil -experimental-skip-non-inlinable-function-bodies -debug-forbid-typecheck-prefix INLINENOTYPECHECK %s -o %t/Skip.noninlinable.sil
+// RUN: %target-swift-frontend -emit-sil -emit-sorted-sil -O -experimental-skip-all-function-bodies -debug-forbid-typecheck-prefix ALLNOTYPECHECK %s -o %t/Skip.all.sil
```
Comment (xymus) [inline test/Frontend/skip-function-bodies.swift:17]: `-debug-forbid-typecheck-prefix` is a nice addition to that test, I didn't know we had this flag.

File: include/swift/Basic/FunctionBodySkipping.h [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+#ifndef SWIFT_BASIC_FUNCTIONBODYSKIPPING_H
+#define SWIFT_BASIC_FUNCTIONBODYSKIPPING_H
+
+#include "llvm/Support/DataTypes.h"
+
+namespace swift {
+
+/// Describes the function bodies that can be skipped in type-checking.
+enum class FunctionBodySkipping : uint8_t {
```
Comment (xymus) [inline include/swift/Basic/FunctionBodySkipping.h:21]: The enum will be useful. I should be adding a new case to it soon.

**Review by xymus — APPROVED** [review]

**Comment by bnbarham** [comment]
@swift-ci please test Linux platform

**Comment by bnbarham** [comment]
Okay... well, I'm going to leave the getSourceRange change out for now. It was previously needed but isn't with the current changes, so it should have just been a small cleanup (apparently not though). The only case I could see it being different is when a Synthesized accessor has a start location, but when would that be the case?

I'll collapse the two commits and re-run the tests.

**Comment by bnbarham** [comment]
@swift-ci please smoke test

File: test/Frontend/skip-function-bodies.swift (hunk 2 above)
Comment (bnbarham) [reply test/Frontend/skip-function-bodies.swift:17]: Heh, the "I didn't know" is me almost every day at the moment :D!

File: include/swift/Basic/FunctionBodySkipping.h (hunk 3 above)
Comment (bnbarham) [reply include/swift/Basic/FunctionBodySkipping.h:21]: Woo, more skipping!

**Comment by bnbarham** [comment]
@swift-ci please smoke test OS X platform

**Comment by bnbarham** [comment]
Going to wait for #34569 to be merged. I'll then rebase and add those new tests into the new combined file.

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please smoke test

**Comment by bnbarham** [comment]
Forgot to run clang-format, but otherwise the same.

**Comment by bnbarham** [comment]
Windows error is unrelated (compile error in llvm-project) and it had passed previous to the clang-format, so just going to merge.

**Pull request merged**